diffedit: add --restore-descendants flag

samueltardieu opened the pull request:
This adds to `diffedit` something that `unsquash` could do and `squash` could not. This would ease the deprecation of `unsquash` (see #4479).

# Checklist

If applicable:
- [X] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [X] I have added tests to cover my changes

martinvonz commented:
See #1027 for some discussion about how we may want to make this more generic. I don't think we should have to duplicate implementation in each command where it's useful.

samueltardieu commented:
> I don't think we should have to duplicate implementation in each command where it's useful.

I have moved the reparenting code into `jj_lib::repo`, and simplified it a lot, thanks to `Repo::transform_descendants()`. Now, the part that is specific to `diffedit` is reduced to calling either `rebase_children` or `reparent_children`.

PhilipMetzger commented on cli/src/commands/diffedit.rs:
```diff
@@ -59,6 +59,9 @@ pub(crate) struct DiffeditArgs {
     /// Specify diff editor to be used
     #[arg(long, value_name = "NAME")]
     tool: Option<String>,
+    /// Do not modify the content of the revision specified with `-r` or `--to`.
```
Nit: No trailing period in doc strings.

samueltardieu replied:
Thanks. I've also cleaned up the rest of the help strings for `diffedit` in #4497.

yuja commented on cli/src/commands/diffedit.rs:
```diff
@@ -59,6 +59,9 @@ pub(crate) struct DiffeditArgs {
     tool: Option<String>,
+    /// Do not modify the content of the revision specified with `-r` or `--to`
+    #[arg(long)]
+    preserve_descendants: bool,
```
nit: `--preserve-descendants` sounds like the default behavior is to somehow destroy descendants. `--reparent-descendants` or `--preserve-descendant-snapshots` ("snapshots" compared to "diffs") might be slightly better.

samueltardieu replied:
I feel like `--preserve-descendant-snapshots` is too long to be usable and too technical. Even `content` would be clearer than `snapshots` for the end-user I think.

martinvonz replied:
I agree that `--preserve-descendant-snapshots` is too technical (not sure I think it's too long because users will probably set up shell completion). I also agree with Yuya's concern. I'm not sure what's a better name. Another way of viewing it is that we apply the reverse change in the descendants, but I suspect it would be even more confusing to explain it that way (`--undo/backout-in-descendants`). Maybe `--restore-descendants` would work? The idea is that it matches the `jj restore` semantics.

samueltardieu replied:
Another point I had read was the `--restore-descendants` could not be later abbreviated as `-r` or `-R` since this is used already, but nothing prevents us from using another letter, that is not a rare practice.

I'll update the PR with `--restore-descendants` and see how it feels.

yuja commented on cli/src/commands/diffedit.rs:
```diff
@@ -125,6 +132,6 @@ don't make any changes, then the operation will be aborted.",
             }
```
nit: "Reparented .."?

samueltardieu replied:
I'm not sure I understand what you're suggesting here, the suggestion seems to be on the "the operation will be aborted" line.

yuja replied:
```rust
if args.preserve_descendants {
    writeln!(formatter, "Reparented ...")
} else {
    writeln!(formatter, "Rebased ...")
}
```
(no idea why I can't add comment to unchanged line on GitHub.)

samueltardieu replied:
Let me try this. I'm not sure this level of details is warranted, but it might be. I'll add this as a separate commit to test the idea for the time being.

samueltardieu replied:
"reorganize"?

samueltardieu replied:
Martin has suggested "rewrite" on Discord, I'll give it a try.

yuja commented on cli/src/commands/diffedit.rs:
```diff
@@ -125,6 +132,6 @@ don't make any changes, then the operation will be aborted.",
         }
         tx.finish(ui, format!("edit commit {}", target_commit.id().hex()))?;
-    }
+    };
```
nit: redundant `;`

samueltardieu replied:
Ooops. Wasn't caught by clippy, I don't even see an applicable lint, I'll investigate.

martinvonz commented on lib/src/repo.rs:
```diff
@@ -1329,6 +1329,21 @@ impl MutableRepo {
         Ok(num_rebased)
     }
 
+    pub fn reparent_descendants(&mut self, settings: &UserSettings) -> BackendResult<usize> {
```
Can you add a doc comment?

Sorry that I forgot to ask before, and sorry that `rebase_descendants()` (which was probably added by me) doesn't have it. I hope to some day find time to go through and add missing doc comments so we can remove `#![allow(missing_docs)]` everywhere.

samueltardieu replied:
I've added an initial commit with documentation for `rebase_descendants()`, and added the documentation for `reparent_descendants()`.

martinvonz commented on cli/src/commands/diffedit.rs:
```diff
@@ -64,6 +64,9 @@ pub(crate) struct DiffeditArgs {
     /// Specify diff editor to be used
     #[arg(long, value_name = "NAME")]
     tool: Option<String>,
+    /// Do not modify the content of the revision specified with `-r` or `--to`
```
Let's try to clarify this. I'm not sure most readers will understand what it means. (It also seems incorrect in that it says that the target commit itself is unchanged). Maybe something like this:

> Preserve the content (not diff) when rebasing descendants
> 
> When rebasing a descendant on top of the rewritten revision, its diff compared to its parent(s) is normally
> preserved, i.e. the same way that descendants are always rebased. This flag makes it so the content/state
> is preserved instead of preserving the diff.

Do you think that's clear?

samueltardieu replied:
The "(not diff)" looked a bit strange to me at first glance, but with the full explanation it makes sense.

martinvonz replied:
not sure if it actually helps, but "(not the diff)" sounds better upon rereading it

samueltardieu replied:
Yes a bit indeed. Updated.

martinvonz commented on cli/src/commands/diffedit.rs:
```diff
@@ -119,13 +122,17 @@ don't make any changes, then the operation will be aborted.",
-            if num_rebased > 0 {
-                writeln!(formatter, "Rebased {num_rebased} descendant commits")?;
+            if num_rewritten > 0 {
+                writeln!(formatter, "Rewritten {num_rewritten} descendant commits")?;
```
nit: Rewritten -> Rewrote

Also - and I know we talked about this on Discord, but I didn't think about it until now - I'm worried that users who don't pass `--restore-descendants` will be confused that this command says "Rewrote 5 descendant commits" when all other commands say "Rebased 5 descendant commits". I'm leaning towards  using "Rebased" for both cases or maybe "Rebased 5 descendant commits" and "Rebased 5 descendant commits (while preserving their content)" or something like that.

samueltardieu replied:
I've added your proposed phrasing. If we add this option to more commands, it will make sense to also factor out the message printing in order not to repeat the same text (and code) everywhere.

samueltardieu replied:
@martinvonz What about using "Rebased 5 descendant commits (no content change)" instead of "Rebased 5 descendant commits (while preserving their content)"?

martinvonz replied:
I think that sounds a bit more like something that happened to be the case given some circumstance than something that's guaranteed to be the case. Perhaps the user will think that it was something about those commits that made it so there was no content change. Does that make sense?

samueltardieu replied:
It does! Leaving as-is.

martinvonz commented on cli/tests/test_diffedit_command.rs:
```diff
@@ -514,3 +516,119 @@ fn test_diffedit_old_restore_interactive_tests() {
+}
+
+#[test]
+fn test_diffedit_restore_descendants_with_multiple_parents() {
```
This test should ideally go into the lib crate, testing the new function in the first commit in this PR. It makes sense to also have a test of the CLI feature, but we should not have to test more than just the basics here. The lib tests are faster, and they mean that we test the reusable functionality, so we don't have to test it for every command that gains a `--restore-descendants` flag. Sorry that I didn't notice this earlier.

samueltardieu replied:
I won't be able to do it right away, but I'll keep this on my list.

martinvonz replied:
No problem. If you think you'll get to it soon, please update the PR. If it might take a bit longer (than a few days, says), feel free to send another PR instead.

samueltardieu replied:
I've added a new test in its own commit

martinvonz replied:
Thanks! I think you can delete this test now, unless you think it covers some case that the jj-lib test and the test above don't cover.

martinvonz reviewed (approved):
LGTM, but let's give Yuya a chance to comment on the option name (and the code too).

martinvonz commented on lib/tests/test_mut_repo.rs:
```diff
@@ -676,3 +676,75 @@ fn test_remove_wc_commit_previous_discardable() {
 }
+
+#[test]
+fn test_reparent_descendants() {
```
nit: squash the test into the parent commit (see https://martinvonz.github.io/jj/latest/contributing/#code-reviews)

yuja commented on lib/tests/test_mut_repo.rs:
```diff
@@ -676,3 +676,76 @@ fn test_remove_wc_commit_previous_discardable() {
+    let empty_commit_tree_id = mut_repo.store().empty_merged_tree_id().clone();
+    mut_repo
+        .rewrite_commit(&settings, &commit_a)
+        .set_tree_id(empty_commit_tree_id)
```
Maybe better to use another random tree (or file content updates) here.

If a file deleted from both merge sides, the result is deletion. This test passes even if we did `rebase_descendants()`.

samueltardieu replied:
Good catch. Fixed.

samueltardieu commented on lib/src/repo.rs:
```diff
@@ -1314,6 +1314,11 @@ impl MutableRepo {
         result
     }
 
+    /// Reparent descendants of the rewritten commits.
```
Ooops, cut-n-paste went wrong. Fixed, thanks.
